### DISCUSSION TITLE
use of <script type="module"> for compatibility with firebase

### DISCUSSION THREAD

**Discussion by mkaesler** [jspsych/jsPsych#3024] [General]
Hi all, 

I'm interested in using firebase to save data from jspsych experiments. Version 9 of firebase is modular, and the functions for interacting with firebase need to be imported in <script type="module">. The functions imported in modules are not available in the global scope, which means that they can't be called in a script tag before the main experiment. This is annoying, because I need these functions to save the data to firebase! 

My simplest workaround is to just make the experiment script type="module" and move the imports to the start of the experiment script. This does work in that the firebase function are now accessible in the experiment script and the data saves, but I'm wondering if : 
a) the strict mode induced by the script tag will break some parts of jspsych, and 
b) if there is another way to make these firebase functions available without changing main experiment script in to a module? 

Kind regards, 

Matt

**Comment by jodeleeuw** [comment]
Hi @mkaesler,

One solution would be to use a tool like webpack to convert the modules. A useful tool here might be [jspsych-builder](https://github.com/bjoluc/jspsych-builder), which provides all the necessary tooling to bundle your modules into something that any browser can interpret.

**Reply by mkaesler** [reply]
Hi @jodeleeuw, 

Just to follow up on this, are there any problems you foresee with having the jspsych script that runs the experiment being type = "module", other than some workarounds required to debug in browser console due to modules having independent scope? 

If I set the experiment script up this way, I can import the relevant firebase functions and connect to the database at the start of the script, and save the data on_finish. It seems to work fine, the data saves, but I just wonder if there's something I'm missing. 

Kind regards, 

Matt

**Reply by jodeleeuw** [reply]
I think that should work fine!

**Comment by mkaesler** [comment]
Hi Josh,

Thanks for the hint re webpack as a way to bundle and convert, and thanks
for the great library too!

Kind regards,

Matt
